src/mlio/image_reader.cxx

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You
+ * may not use this file except in compliance with the License. A copy of
+ * the License is located at
+ *
+ *      http://aws.amazon.com/apache2.0/
+ *
+ * or in the "license" file accompanying this file. This file is
+ * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+ * ANY KIND, either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ */
+
+#include "mlio/image_reader.h"
+
+#include <fmt/format.h>
+#include <opencv4/opencv2/imgcodecs.hpp>
+#include <opencv4/opencv2/imgproc/imgproc.hpp>
+
+#include "mlio/cpu_array.h"
+#include "mlio/data_type.h"
+#include "mlio/instance.h"
+#include "mlio/instance_batch.h"
+#include "mlio/intrusive_ptr.h"
+#include "mlio/parallel_data_reader.h"
+#include "mlio/record_readers/blob_record_reader.h"
+#include "mlio/record_readers/recordio_record_reader.h"
+#include "mlio/schema.h"
+#include "mlio/streams/utf8_input_stream.h"
+
+namespace mlio {
+inline namespace v1 {
+
+image_reader::image_reader(data_reader_params rdr_prm,
+                           image_reader_params image_prm)
+    : parallel_data_reader{std::move(rdr_prm)}, params_{std::move(image_prm)}
+{
+    if (params_.data_shape.size() != data_shape_size_) {
+        throw std::invalid_argument{
+            "data_shape is a required argument. "
+            "Please enter shape of the output image in "
+            "(channels, height, width) format"};
+    }
+}
+
+image_reader::~image_reader()
+{
+    stop();
+}
+
+intrusive_ptr<record_reader>
+image_reader::make_record_reader(data_store const &ds)
+{
+    if (params_.img_frame != image_frame::file &&
+        params_.img_frame != image_frame::pipe) {
+        throw std::invalid_argument{"Please enter a valid value for "
+                                    "image_frame parameter."};
+    }
+    switch (params_.img_frame) {
+    case image_frame::file:
+        return make_intrusive<blob_record_reader>(ds.open_read());
+    case image_frame::pipe:
+        return make_intrusive<detail::recordio_record_reader>(ds.open_read());
+    }
+}
+
+void
+image_reader::infer_schema(instance const &)
+{
+    std::vector<feature_desc> descs{};
+    // schema follows the NHWC convention
+    descs.emplace_back(
+        feature_desc_builder{"value",
+                             data_type::uint8,
+                             {params().batch_size,
+                              static_cast<std::size_t>(params_.data_shape[1]),
```

Comment:
So what I noticed was that making `data_shape` of type `std::size_t` lead to more typecasting to int at various places since opencv `Mat` used `int` for matrix rows/cols. This way, I was able to achieve cleaner code as number of typecasting decreased?

thoughts?

Reply:
I was referring to this as the earlier comment.